### ISSUE TITLE
Consider generating multiple bundles in favor of modern browsers

### ISSUE THREAD

**Issue by kripod** [gatsbyjs/gatsby#2114]
According to [this article](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/), modern browsers may profit from bundles loaded with support for the ES2015 module syntax. Using the methods described in the article may also reduce server load because of the reduced bundle sizes.

**Comment by KyleAMathews** [comment]
Huh super interesting! Would love to see someone set this up on gatsby! Should be doable as a plugin.

Btw, the main benefit appears to be that you don't have to polyfill for newer browsers so the code loaded is significantly smaller.

**Comment by jquense** [comment]
That article seems confused (or I am)...the main jist is to not compile modules via babel. The problem with that tho is that you don't get a bundle out that has uncompiled es modules, you get a compiled webpack bundle with _no_ modules. Babel won't compile the modules but webpack *will*. So you can use the "modern" babel config with any browser, in fact the v2 gatsby branch (which uses webpack2) already does this.

**Comment by jquense** [comment]
ahhhh I get it sorry i'm the dense one. The point is just that you can simplify feature detection by checking `script module` tags?

**Comment by KyleAMathews** [comment]
@davidtheclark over at Batfish is working on a PR for a similar feature https://github.com/mapbox/batfish/pull/188

This technique seems *really* promising.

We'd need to see how much time it adds to the build but if it's not that much more, I'd definitely consider turning it on by default. If the extra build time hurts enough, we could make it optional.

**Comment by robwierzbowski** [comment]
@jquense I'm really into this approach. To summarize my understanding:

- Transpiled code is often larger and less performant than es6+ code.
- `script type="module"` is a good "cut the mustard" line for a lot of es6+ features. Remember [cut the mustard](http://responsivenews.co.uk/post/18948466399/cutting-the-mustard) from maybe 5-6 years ago?
- We should still bundle our modules
- We can create a bundle with much less transpiling for modern browsers, and use  `script type="module"` as a sneaky feature detect to load a minimally transpiled bundle on newer browsers, and a more transpiled / es5 bundle on older browsers.

**Comment by KyleAMathews** [comment]
Yup! And per our parallel polyfill discussion, the  `script type="module"` would need pretty minimal polyfills and be very cheap to deliver them over CDN.

**Comment by davidtheclark** [comment]
I ended up holding off on that Batfish PR but implemented the plan on another build system tool — with the idea of bringing it into Batfish after Babel 7 is released. It worked well overall, but there were two hiccups:

- Minification becomes more complicated, because you want to minify your ES2015+ bundle without having to compile it to ES5. Uglify has been moving pretty fast to catch up to ES2015+, though, so this might not be much of an issue anymore. Seems to be working ok for me.
- You have to run the Webpack build twice, which is kind of a bummer.

**Comment by robwierzbowski** [comment]
Yeah, the build twice is a pain. Webpack persistant caching might hep that, depending on how it’s implemented. 

Babel-minify worked well for me too. Not quite as small as uglify es branch, but not as many edge cases either.

**Comment by mquandalle** [comment]
For reference Meteor 1.7 introduced the ability to generate two different JavaScript bundles, one for legacy browsers that includes JS polyfills for features that aren't natively implemented by the browser, and one for modern browsers without these polyfills. The rationale for the Meteor design is exposed in https://github.com/meteor/meteor/pull/9439 and https://blog.meteor.com/meteor-1-7-and-the-evergreen-dream-a8c1270b0901.

**Comment by benjamn** [comment]
@KyleAMathews I'd be happy to discuss how this might work in Gatsby! It's a nice selling point for Meteor right now, but I think it's safe to say Gatsby's users have different needs, and are not about to switch to Meteor.

Most of my thoughts are in the blog post that @mquandalle linked above, but here's a slightly shorter explanation that I recently typed in a FB comment:

>  Here's the gist: In previous versions of Meteor, we had only one set of static assets (JS, CSS, etc.), and those assets needed to work in any browser that we claimed to support.

> In Meteor 1.7, these assets have become "legacy," which means they still work perfectly well in any browser that we support. Of course, in order to support older browsers, the JS bundle needed to be compiled using Babel, and the resulting code was larger, messier, less native.

> In Meteor 1.7, we wanted to start shipping a smaller, less messy, more native bundle tailored for "modern" browsers that have native support for a growing set of ECMAScript features. To do that, we considered each polyfill and Babel plugin that we were previously using, and asked what would happen if we just stopped using that polyfill or plugin. Of course, removing a plugin that handles class syntax means your bundle can only run in browsers that natively support class syntax, so there's a cost to removing each plugin. Removing polyfills and plugins narrows down the set of browsers that can be considered "modern" (which just means it's safe to send them whatever modern bundle we built).

> So that's essentially where the modern bundle comes from: it's a copy of the legacy bundle, with as many polyfills and plugins *removed* as we felt comfortable removing. For every feature that we stopped simulating in older browsers, we found the minimum version of each browser that natively supports that feature, so only those browser versions are eligible to receive the modern bundle. More precisely, to receive the modern bundle, a browser needs to meet the minimum version for *all* the features that we no longer compile.

> You might think that removing, say, half of the plugins would result in a bundle that only a very small percentage of browsers could handle, but the amazing thing is that more than 80% of the world's browsers are "evergreen," and they all support basically the whole ECMAScript standard, so you have to remove a whole lot of plugins before you end up with a bundle that they can't handle. We'd be in a very different (worse) situation if every major browser supported some arbitrary partial set of ECMAScript features. Luckily, the evergreen browsers are in vigorous agreement.

> Right now, the most bleeding-edge feature that we leave uncompiled in the modern bundle is probably async/await. Yes, that was just finalized in 2017, but all the evergreen browsers already support it. At some point we'll be able to stop compiling things like decorators and public/private class properties, because enough browsers will support those features natively, but we're not there yet.

> Sniffing user agents can be dicey, but if the server is ever not 100% confident about the browser/version of a client, we can always fall back to the legacy bundle.

**Comment by benjamn** [comment]
One quick question: does Gatsby have any mechanism for delivering different HTML to different browsers, or is the `<script type="module">`/`<script nomodule>` (for differentiating between JS bundles) the only trick a static site can use?

**Comment by m-allanson** [comment]
Gatsby probably needs to rely on `<script type="module">/<script nomodule>`

**Comment by benjamn** [comment]
I think that's a perfectly reasonable strategy, though I just finished tweeting some thoughts on the limits of that approach: https://twitter.com/benjamn/status/1005104014944538624

**Comment by KyleAMathews** [comment]
Vue is shipping with this as an optional mode for builds https://cli.vuejs.org/guide/browser-compatibility.html#modern-mode

**Comment by alexparish** [comment]
This would be a fantastic addition for v2, but I understand there are many competing priorities for that release.

Out of interest does anyone know if the non-transpiled ES6 (eg. arrow functions, classes) executes as fast as transpiled code in modern browsers now?

**Comment by KyleAMathews** [comment]
Not stale!

**Comment by KyleAMathews** [comment]
This is a through guide to setting this up as well https://calendar.perfplanet.com/2018/doing-differential-serving-in-2019/

**Comment by TylerBarnes** [comment]
A colleague mentioned to me that brotli compression is considerably better than gzip compression and it's supported by all major modern browsers. There is a [webpack plugin](https://www.npmjs.com/package/brotli-webpack-plugin) and I'm wondering if this might be a good addition to the modern browser bundle? According to [this article](http://www.instantshift.com/2018/03/02/gzip-vs-brotli-compression/) it's 14% smaller for js files, 21% smaller for html files, and 17% smaller for css files. This actually the first I've heard of it though so I'm not sure if there is some reason not to do it.

**Comment by KyleAMathews** [comment]
Brotli is great! The trick there is that not all CDNs/hosting providers support it yet — so it's a bit trickier to support.

**Comment by TylerBarnes** [comment]
That makes sense, that is tricky. It would be mindblowing to see gatsby shave off even that much more since you guys are already doing such an incredible job on optimization. Either way I'm excited to see what comes out of this issue.

**Comment by pieh** [comment]
Another resource/guide for setting up webpack + babel to produce 2 sets of bundles https://web.dev/fast/serve-modern-code-to-modern-browsers/codelab-serve-modern-code

**Comment by Ronva** [comment]
@pieh and I have started tackling this and we made some good progress. Turns out this is much more complex than either of us thought.

**Progress**
Using an environment variable, we loaded a modern babel preset env configuration that generated smaller bundles (about 15% smaller uncompressed for the default starter) with an mjs extension. These files weren't loading by default because they were lacking the `type="module"` property, so we modified that where the HTML output is generated and it worked like a charm.

**Roadblocks**
To get this to work we need two babel configs and ultimately generate two separate webpack configs. While duplicating the webpack configuration step created js chunks and mjs chunks, something in the webpack stats file was overwriting one or the other. I need to look further into how Gatsby handles custom babel configs and ask that if anyone knows how this works to offer some insight.

Another thing we noticed is that node_modules was caching the generated files so we had to delete the cache folder for changes in the chunk generation to take effect. This might not be an issue once we take care of that first issue.

**Observations**
Webpack loads runtime chunks (prefetched pages, etc.) using [runtimeChunk](https://webpack.js.org/configuration/optimization/#optimization-runtimechunk) which did load mjs files but didn't add `type="module"` to the script tag since the property is only added to scripts created at build time. This didn't actually affect the loading of javascript but might give warnings in the console. 

My initial concern was that it would read both the modern code and legacy code if webpack were to load mjs scripts that don't have the type property. This actually isn't a concern because the chunk that generates the runtime chunks is created at build time, meaning only the appropriate runtime chunks will be loaded (let me know if this is not explained well, it's a little hard to phrase properly).

Finally, I would like to look into passing the browser list generated by the modern babel config into postcss assuming it meaningfully affects the size of generated CSS. This would require researching the effect autoprefixer has on file size. This is definitely low priority.

**Comment by KyleAMathews** [comment]
Why a separate babel config? I thought you just had to change the "target" key? Or is that what you mean by "separate"? :-)

Also what are you changing in webpack? Is there a branch or PR w/ your code changes so far?

Why are you generating `.mjs` files instead of `.js` files? The entry file needs to be `.mjs` if `type=module`?

I'd assume the two builds create two webpack runtime modules right? You can set the file name for the stats file so I assume that would solve the overwriting problem?

Thanks for diving into this!

**Comment by pieh** [comment]
> Why a separate babel config? I thought you just had to change the "target" key? Or is that what you mean by "separate"? :-)

It is almost same babel config, just options for `@babel/preset-env` are different.

> Also what are you changing in webpack? Is there a branch or PR w/ your code changes so far?
>
> Why are you generating `.mjs` files instead of `.js` files? The entry file needs to be `.mjs` if `type=module`?

For the most part just `output` field to produce files with `.mjs` ext - I guess extension itself might not matter - we were replicating guides here on that - might be not needed.

> I'd assume the two builds create two webpack runtime modules right? 

It will, and those should have either `type="module"` or `nomodule` on their script tag, so only one of them will be fetched and loaded

> You can set the file name for the stats file so I assume that would solve the overwriting problem?

Yeah, it's not really hard problem - we were just documenting stuff that need to be done.

**Comment by Ronva** [comment]
@KyleAMathews Yup, by "separate" I only mean changing targets to `{ esmodules: true }`, and loose + useBuiltIns to false in babel preset env's options. [Changes found here](https://github.com/Ronva/gatsby/blob/967915aa9181aac8969b0fd33a679588cfba7231/packages/babel-preset-gatsby/src/index.js#L16-L29)

**Comment by KyleAMathews** [comment]
Cool! This is super exciting!

**Comment by KyleAMathews** [comment]
This is how vue-cli does modern builds https://github.com/vuejs/vue-cli/blob/38efc032e5c0d5ab8dd74602c1fb100fef62d766/packages/%40vue/cli-service/lib/webpack/ModernModePlugin.js

**Comment by davilima6** [comment]
This is article that pulled me into this topic: https://nystudio107.com/blog/an-annotated-webpack-4-config-for-frontend-web-development It doesn't use HtmlWebpackPlugin but it still is a great example of a complete use of multi-target compilations: https://github.com/nystudio107/annotated-webpack-4-config/

**Comment by m-allanson** [comment]
@KyleAMathews and @DSchau you were working on this recently I think? Could you post an update of where you got to? Was there some issues with handling multiple webpack configs?

**Comment by DSchau** [comment]
@m-allanson thanks for tagging us in!

We did. Here's where we ended up (with more work on this probably coming this week):

The current state of the work is [here](https://github.com/DSchau/gatsby/pull/7).

What we'll need to do is do two webpack builds (ideally in parallel, utilizing something like [jest-worker](https://www.npmjs.com/package/jest-worker)), one with a modern config, and one with the legacy config.

Our approach is _tentatively_ going to be opt-out, e.g. `gatsby build --legacy`, because we can't really see (currently!) why a dev wouldn't want this behavior.

Current blockers/issues/things to solve:

- Parallelizing build; we'll probably want to do this for performance reasons
    - See [Kyle's tweet](https://twitter.com/kylemathews/status/1099006785111842816) for some more info here
    - We _can_ build both in the same process, but it's possible this may introduce significantly slower builds for people, so we'll want to weigh the cost of this
- Update the `static-entry.js` file in cache-dir to support `module` `nomodule` script tags
    - Additionally check out the [Vue Safari fix and implementation](https://github.com/vuejs/vue-cli/blob/38efc032e5c0d5ab8dd74602c1fb100fef62d766/packages/%40vue/cli-service/lib/webpack/ModernModePlugin.js)
- Testing, testing, testing
    - We'll want to profile this, validate it works as expected, minimal build performance regression (or none at all), etc. etc.

Additional things that should probably be solved:

- Transpiling our dependencies (e.g. gatsby-image, gatsby-link)
    - We currently publish these as CommonJS - to get some more savings, we should publish a modern build of this (using microbundle?) and then these will be minimally transpiled with the modern build, and then certain features will be transpiled for the legacy build
    - More broadly speaking, could be nice to start transpiling node_modules as a whole, but that's a separate conversation

I think this is a decent intro on the current state of things. More to come soon!

**Comment by Ronva** [comment]
Hey, sorry about the radio silence. A bunch of stuff came up resulting in me not being to continue working on this feature. Still really hoping to see this get done, please let me know of there's anything I can help with!

**Comment by DSchau** [comment]
See: gatsbyjs/gatsby#12761 for a WIP PR @wardpeet and I worked on to use microbundle to publish our browser dependencies (e.g. gatsby-image, gatsby-link) as es and cjs bundles.

We're waiting for some upstream changes to microbundle, then we can merge those in!

**Comment by wardpeet** [comment]
slow making microbundle ready for modern builds
https://github.com/developit/microbundle/pull/362/

**Comment by codepunkt** [comment]
Is there a status update on this?

**Comment by KyleAMathews** [comment]
more links

- https://github.com/zeit/next.js/issues/7563#issuecomment-509823285
- https://github.com/babel/preset-modules/blob/master/README.md

**Comment by jquense** [comment]
Has anyone identified even the correct pattern for doing this in webpack? I've wanted to set this up multiple times but just completely unsure waht the Right Way to even integrate it into webpack is

**Comment by developit** [comment]
@jquense If you're already running a MultiCompiler, the best solution is to add an additional top-level compiler pass. Otherwise, [babel-esm-plugin](https://github.com/prateekbh/babel-esm-plugin) performs the `module` build using a Child Compiler, which is runs in parallel to the parent compilation to a limited extent.

The problem with all current approaches is that they don't extend the benefits of modern JS to code obtained from node_modules. Thankfully, there has recently been some solid movement towards a standardized mechanism for exposing modern JS in npm packages, and I'm hoping to have a demonstration of this soon. It will take a while for packages to adopt the approach, but shipping it as a default in microbundle and tsdx can serve as a nice head-start.
